File: GFG/LinkedList/detectloop.cpp
#include<iostream>
#include"../ds.h"
using namespace std;

Node<int>* loopnode(LinkedList<int> &dl){
    Node<int>* fcurr=dl.head->next->next,*scurr=dl.head->next;
    while(fcurr!=scurr){
        scurr=scurr->next;
        fcurr=fcurr->next->next;
    }
    scurr=dl.head;
    while(true){
        if(fcurr==scurr){
            return fcurr;
        }
        fcurr=fcurr->next;
        scurr=scurr->next;
    }
}

int main(){
    LinkedList<int> dll;
    for(int i=1;i<6;i++){
        dll.insert(i*5);
    }
    Node<int>* temp=new Node(500);
    dll.tail->next=temp;
    dll.tail=dll.tail->next;

    for(int i=6;i<11;i++){
        dll.insert(i*10);
    }
    dll.tail->next=temp;
    dll.tail=dll.tail->next;
    cout<<endl;
    int i=0;
    Node<int>*res=loopnode(dll);
    cout<<"found loopnode  -  ";
    cout<<res->data<<" "<<res->next->data<<endl;
    cout<<"**********"<<endl;
    return 0;
}

File: GFG/LinkedList/detectloop.py
import ds

def loopnode(dl):
    fcurr=dl.head.next.next
    scurr=dl.head.next
    while(scurr!=fcurr):
        scurr=scurr.next
        fcurr=fcurr.next.next
    scurr=dl.head
    while(True):
        if(scurr==fcurr):
            return fcurr
        scurr=scurr.next
        fcurr=fcurr.next

dll=ds.LinkedList()
dll.insert(1)
tail=dll.head
for i in range(1,6):
    dll.insert(i*5)
    tail=tail.next
temp=ds.Node(500)
tail.next=temp
tail=tail.next
for i in range(6,11):
    dll.insert(i*10)
    tail=tail.next
tail.next=temp
i=0
curr=dll.head
while(curr!=None and i<20):
    print(curr.data)
    curr=curr.next
    i+=1
print(tail.data," ",tail.next.data)
print("**********")
print(loopnode(dll).data)